### PR TITLE
cart: Create PaymentSelection and evaluate GiftCard

### PR DESCRIPTION
**Pull request by danielpoe**
If a card has applied giftcards, it means that the card should not by payed in total with the ChargeTypeMain (because some value is payed with a giftcard)

So we need a standard method that returns a paymentselection that takes the giftcard into account

### CODE REVIEW

File: cart/domain/cart/paymentselection.go [hunk 1]
```diff
@@ -87,6 +87,93 @@ func NewSimplePaymentSelection(gateway string, method string, pricedItems Priced
...
+	ratio float64, totalGCValue price.Price) (*PaymentSplitByItemBuilder, price.Price, error) {
+	var remainingItemValue, appliedGcAmount price.Price
+	var err error
+	// loop over given prices and calculate relative amounts
+	for k, itemPrice := range prices {
+		remainingItemValue, totalGCValue, appliedGcAmount, err = calcRelativeGiftcardAmount(itemPrice, totalGCValue, ratio)
+		if err != nil {
+			return nil, totalGCValue, err
+		}
+		builder.AddTotalItem(k, method, price.Charge{
```
[the remark this answers is not in the document]
Comment (jeinfeldt) [reply cart/domain/cart/paymentselection.go]: Correct and adjusted, all items are added now according to their type

File: cart/domain/cart/paymentselection.go [hunk 2]
```diff
@@ -87,6 +87,93 @@ func NewSimplePaymentSelection(gateway string, method string, pricedItems Priced
...
+	if totalGCValue.IsZero() {
+		return NewSimplePaymentSelection(gateway, method, pricedItems), nil
+	}
+	if totalGCValue.IsGreaterThen(totalValue) {
+		return nil, errors.New("giftcard amount exceeds total priced items value")
+	}
+	// distribute gift card amounts relatively across all items
+	giftCartAmountRatio := totalGCValue.FloatAmount() / totalValue.FloatAmount()
+	builder := &PaymentSplitByItemBuilder{}
+	prices := []map[string]price.Price{pricedItems.CartItems(), pricedItems.ShippingItems(), pricedItems.TotalItems()}
```
[the remark this answers is not in the document]
Comment (jeinfeldt) [reply cart/domain/cart/paymentselection.go]: CartItems are Added to the PaymentSplit via addCartItem

File: cart/domain/cart/paymentselection_wb_test.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+			Applied: domain.NewFromInt(100, 100, "€"),
+		},
+		{
+			Applied: domain.NewFromInt(200, 100, "€"),
+		},
+	}
+
+	selection, err := NewPaymentSelectionWithGiftCard("gateyway", "method", pricedItems, appliedGc)
+	assert.NoError(t, err)
+	assert.Equal(t, domain.NewFromInt(1198, 100, "€").FloatAmount(), selection.TotalValue().FloatAmount())
```
[the remark this answers is not in the document]
Comment (jeinfeldt) [reply cart/domain/cart/paymentselection_wb_test.go]: Tests have been adjusted to test the actual totals

File: cart/domain/cart/paymentselection_wb_test.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+			Applied: domain.NewFromInt(5, 1, "€"),
+		},
+	}
+	selection, err := NewPaymentSelectionWithGiftCard("gateyway", "method", pricedItems, appliedGc)
+	assert.NoError(t, err)
+	// verfiy complete cart splits
+	assert.Equal(t, domain.NewFromInt(10, 1, "€").FloatAmount(), selection.CartSplit().ChargesByType().GetByTypeForced(ChargeTypeGiftCard).Value.FloatAmount())
+	assert.Equal(t, domain.NewFromInt(0, 1, "€").FloatAmount(), selection.CartSplit().ChargesByType().GetByTypeForced(domain.ChargeTypeMain).Value.FloatAmount())
+	// verify first product charges
+	relativeGCValue := selection.ItemSplit().TotalItems["1"].ChargesByType().GetByTypeForced(ChargeTypeGiftCard)
```
[the remark this answers is not in the document]
Comment (jeinfeldt) [reply cart/domain/cart/paymentselection_wb_test.go]: Has been adjusted

File: cart/domain/cart/giftcard.go [hunk 5]
```diff
@@ -22,6 +23,28 @@ type (
...
+
+	// PaymentSplitWithGiftCards handles payment split on item level including gift cards
+	PaymentSplitWithGiftCards struct {
+		Method      string
+		PricedItems PricedItems
+		GiftCards   AppliedGiftCards
+	}
+
+	// ErrSplitNoGiftCards indicates that there are no gift cards given to PaymentSplitWithGiftCards
+	ErrSplitNoGiftCards struct {
```
[the remark this answers is not in the document]
Comment (danielpoe) [reply cart/domain/cart/giftcard.go]: if err == ErrSplitNoGiftCards

File: cart/domain/cart/paymentselection.go [hunk 6]
```diff
@@ -87,6 +96,34 @@ func NewSimplePaymentSelection(gateway string, method string, pricedItems Priced
 	return selection
 }
 
+//NewPaymentSelectionWithGiftCard - returns Selection with given giftcard charge type taken into account
+func NewPaymentSelectionWithGiftCard(gateway string, method string, pricedItems PricedItems, appliedGiftCards []AppliedGiftCard) (PaymentSelection, error) {
```
Comment (danielpoe) [inline cart/domain/cart/paymentselection.go]: Maybe we should refactor this to only expose one NewDefaultPaymentSelection(gateway,method,cart) PaymentSelection, error

That makes the public interface easier to use and understand

**Pull request merged**